libsolidity/lsp/LanguageServer.cpp

```diff
@@ -400,3 +426,89 @@ void LanguageServer::handleTextDocumentDidClose(MessageID _id, Json::Value const
 }
+
+ASTNode const* LanguageServer::requestASTNode(std::string const& _sourceUnitName, LineColumn const& _filePos)
+{
+	if (m_compilerStack.state() < CompilerStack::AnalysisPerformed)
+		return nullptr;
+
+	if (!m_fileRepository.sourceUnits().count(_sourceUnitName))
+		return nullptr;
+
+	optional<int> sourcePos = m_compilerStack.charStream(_sourceUnitName)
+		.translateLineColumnToPosition(_filePos);
+	if (!sourcePos.has_value())
+		return nullptr;
+
+	return locateInnermostASTNode(*sourcePos, m_compilerStack.ast(_sourceUnitName));
+}
+
+void LanguageServer::handleGotoDefinition(MessageID _id, Json::Value const& _args)
+{
+	string const uri = _args["textDocument"]["uri"].asString();
+	string const sourceUnitName = m_fileRepository.clientPathToSourceUnitName(uri);
+	if (!m_fileRepository.sourceUnits().count(sourceUnitName))
+	{
+		m_client.error(_id, ErrorCode::RequestFailed, "Unknown file: " + uri);
+		return;
+	}
+
+	auto const lineColumn = parseLineColumn(_args["position"]);
+	if (!lineColumn)
+	{
+		m_client.error(
+			_id,
+			ErrorCode::RequestFailed,
+			fmt::format(
+				"Unknown position {line}:{column} in file: {file}",
+				fmt::arg("line", lineColumn.value().line),
+				fmt::arg("column", lineColumn.value().column),
+				fmt::arg("file", sourceUnitName)
+			)
+		);
+		return;
+	}
+
+	ASTNode const* sourceNode = requestASTNode(sourceUnitName, lineColumn.value());
+	vector<SourceLocation> locations;
+	if (auto const* identifier = dynamic_cast<Identifier const*>(sourceNode))
+	{
+		for (auto const* declaration: allAnnotatedDeclarations(identifier))
+			if (auto location = declarationPosition(declaration); location.has_value())
+				locations.emplace_back(move(location.value()));
+	}
+	else if (auto const* identifierPath = dynamic_cast<IdentifierPath const*>(sourceNode))
+	{
+		if (auto const* declaration = identifierPath->annotation().referencedDeclaration)
+			if (auto location = declarationPosition(declaration); location.has_value())
+				locations.emplace_back(move(location.value()));
+	}
+	else if (auto const* memberAccess = dynamic_cast<MemberAccess const*>(sourceNode))
+	{
+		auto const location = declarationPosition(memberAccess->annotation().referencedDeclaration);
+		if (location.has_value())
+			locations.emplace_back(location.value());
+	}
+	else if (auto const* importDirective = dynamic_cast<ImportDirective const*>(sourceNode))
+	{
+		auto const& path = *importDirective->annotation().absolutePath;
+		if (m_fileRepository.sourceUnits().count(path))
+			locations.emplace_back(SourceLocation{0, 0, make_shared<string const>(path)});
+	}
+	else if (auto const* declaration = dynamic_cast<Declaration const*>(sourceNode))
+	{
+		if (auto location = declarationPosition(declaration); location.has_value())
+			locations.emplace_back(move(location.value()));
+	}
+	else if (sourceNode)
+	{
+		m_client.error(_id, ErrorCode::InternalError, fmt::format("Could not infer def of {}", typeid(*sourceNode).name()));
```

Comment:
Shouldn't we just return an empty array in this case? `null` also seems to be an option.

Reply:
@chriseth yes. But I want to log that case somehow, and we do not have a logging facility anymore. I wonder what might be the best that you'd be willing to merge. ;-)

Reply:
`ofstream("/tmp/log.txt", ofstream::app) << "My message." << endl;`

Reply:
This is however not being sent to the editor client. I will rewrite this line however.